app/services/induction/delete_dup_record.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+class Induction::DeleteDupRecord < BaseService
+  COMPARE_ATTRIBUTES = %(induction_programme_id participant_profile_id schedule_id training_status preferred_identity_id school_transfer appropriate_body_id)
+
+  class DeleteInductionRecordRestrictionError < StandardError; end
+
+  def call
+    ActiveRecord::Base.transaction do
+      check_record_is_deletable!
+
+      update_previous_record
+      delete_induction_record
+    end
+  end
+
+private
+
+  attr_reader :induction_record
+
+  def initialize(induction_record:)
+    @induction_record = induction_record
+  end
+
+  def check_record_is_deletable!
+    raise DeleteInductionRecordRestrictionError, "Cannot delete record because it is active" if active?
+    raise DeleteInductionRecordRestrictionError, "Cannot delete record because it is not in the middle of the induction records history" unless middle_of_history?
+    raise DeleteInductionRecordRestrictionError, "Cannot delete record because it has been diverted from the previous record" if record_changed?
+    raise DeleteInductionRecordRestrictionError, "Cannot delete record because the mentor does not matches the previous record" unless valid_mentor_value?
+  end
+
+  def active?
+    induction_record.active?
+  end
+
+  def middle_of_history?
+    previous_record && next_record
+  end
+
+  def record_changed?
+    induction_record.attributes.slice(*COMPARE_ATTRIBUTES) != previous_record.attributes.slice(*COMPARE_ATTRIBUTES)
+  end
+
+  def valid_mentor_value?
+    induction_record.mentor_profile_id.nil?
+  end
+
+  def delete_induction_record
+    induction_record.destroy
+  end
+
+  def update_previous_record
+    previous_record.update!(end_date: next_record.start_date)
+  end
+
+  def previous_record
+    participant_induction_records[induction_record_index + 1] unless last_record?
+  end
+
+  def next_record
+    participant_induction_records[induction_record_index - 1] unless first_record?
+  end
+
+  def participant_induction_records
+    @participant_induction_records ||= induction_record
```

Comment:
If for every single IR to delete we need to find their participant and all their IRs, this could explain why it is taking so long to finish execution.

Is it too late for a solution where you iterate in batches over the participant_profiles affected by this issue  rather than IR by IR? This would allow to delete at once all the IRs of a participant

Reply:
I'm on the support rota, so not sure if I have time this week.

Wanna give it a try?

Reply:
By the way, we decided to delete the records one by one as it was the safest way to ensure the IR sequence will be maintained.